ark/schema/structure/structure.ts

```diff
@@ -299,6 +324,40 @@ const implementation: nodeImplementationOf<Structure.Declaration> =
 
 				return childIntersectionResult
 			}
+		},
+		reduce: (inner, $) => {
+			if (inner.index) {
+				let hasUpdated = false
+
+				const requiredProps = inner.required ?? []
+				const optionalProps = inner.optional ?? []
+				const newOptionalProps: OptionalNode[] = [...optionalProps]
+
+				if (requiredProps.length === 0 && optionalProps.length === 0) return
+
+				for (const index of inner.index) {
+					for (const requiredProp of requiredProps) {
```

Comment:
I think we can simplify this a bit if we:

1. iterate over a single list of both required and optional props
2. basically just use the optional loop logic there

We shouldn't even have to check if we're dealing with a required or optional prop since optional is the only one that can return a new node.

Reply:
I did consider taking this approach beforehand, but ultimately decided this was better readability. As someone that's not super familiar with the intricacies of the codebase, having two separate loops makes things clearer.